### DISCUSSION TITLE
How to test _app.tsx with Jest

### DISCUSSION THREAD

**Discussion by lauragift21** [vercel/next.js#13717] [Help] [answered]
I created a custom `_app.tsx` file so I can import the tailwindcss base file.

```
import '../styles/index.css'

export default function MyApp({ Component, pageProps }) {
  return <Component {...pageProps} />
}
```

Now trying to write a test to cover the file

```import { render } from '@testing-library/react'
import MyApp from '../../pages/_app'

test('renders snapshot of app component', () => {
  const tree = render(MyApp(<MyApp {...props}  />))
  console.log(tree)
  expect(tree).toMatchSnapshot()
})
```

But I keep getting this typescript error. Anyone know what I'm not doing correctly. It's my first time working with Typescript.
```Type '{}' is missing the following properties from type '{ Component: any; pageProps: any; }': Component, pagePropsts(2739)```

**Comment by jamesmosier** [comment, marked as answer]
Hi @lauragift21! I'm not sure testing `_app.tsx` will provide you with much coverage, as this is mainly just used internally by Next and all it does in this case is renders whatever component Next passes to it (which would be a `page` component).

As far as the error, you will want to add the `AppProps` type to your `_app.tsx` [like this](https://nextjs.org/docs/basic-features/typescript#custom-app):


```jsx
import { AppProps } from 'next/app'
import '../styles/index.css'

function MyApp({ Component, pageProps }: AppProps) {
  return <Component {...pageProps} />
}

export default MyApp
```

**Reply by lauragift21** [reply]
Thank you @jamesmosier

**Reply by hdr-js** [reply]
I understood about the minimal amount of coverage one can get by adding a test for _app. But for sharing some data between pages, I once needed redux, which has a Provider wrapped around the <App>. In that case, I guess, I could've wrote one to test the store inclusion etc.

**Reply by luistak** [reply]
@jamesmosier What if my `_app.tsx` wraps something like a pageview tracking?

Example:
```tsx
import { FC, useEffect } from 'react';
import type { AppProps } from 'next/app';

import * as gtag from 'lib/gtag';

const App: FC<AppProps> = ({ Component, pageProps, router }) => {
  useEffect(() => {
    const handleRouteChange = (url: string) => {
      gtag.pageview(url);
    };
    router.events.on('routeChangeComplete', handleRouteChange);
    return () => {
      router.events.off('routeChangeComplete', handleRouteChange);
    };
  }, [router.events]);

  return <Component {...pageProps} />;
};

export default App;
```

This kind of behavior would be nice to ensure while navigating between pages.

Is there any recommended way to test it?
Or this kind of behavior should be placed in another component? or a custom hook at the `_document.tsx` ?

**Reply by garmeeh** [reply]
Anyone ever figure out if there was recommended way of testing this kinda behaviour?

**Reply by Kshitij978** [reply]
> @jamesmosier What if my `_app.tsx` wraps something like a pageview tracking?
> 
> Example:
> 
> ```tsx
> import { FC, useEffect } from 'react';
> import type { AppProps } from 'next/app';
> 
> import * as gtag from 'lib/gtag';
> 
> const App: FC<AppProps> = ({ Component, pageProps, router }) => {
>   useEffect(() => {
>     const handleRouteChange = (url: string) => {
>       gtag.pageview(url);
>     };
>     router.events.on('routeChangeComplete', handleRouteChange);
>     return () => {
>       router.events.off('routeChangeComplete', handleRouteChange);
>     };
>   }, [router.events]);
> 
>   return <Component {...pageProps} />;
> };
> 
> export default App;
> ```
> 
> This kind of behavior would be nice to ensure while navigating between pages.
> 
> Is there any recommended way to test it? Or this kind of behavior should be placed in another component? or a custom hook at the `_document.tsx` ?

Did you find the answer?

**Reply by dani-vh** [reply]
@jamesmosier I think that the goal of being able to test using `_app` as root component is to ensure that a given flow is working fine from end to end without having to mock things like contexts, which is my case, I'm definning my contexts in the `_app` by wrapping the `Component`, and I'm trying to test from the loging to the page and scenario I need.

Just something like what @luistak need, some jest tests could involve scenarios that includes logic defined inside the `_app`.

**Comment by armikhael** [comment]
Hello @lauragift21. What did you do to test the nextjs component? It gives me the following error. Thanks
![Captura de Pantalla 2022-08-18 a la(s) 00 23 00](https://user-images.githubusercontent.com/4640988/185293224-71d67571-5767-4052-a9e6-a317963fcc5e.png)


```import { render } from "@testing-library/react";
import MyApp from "../../pages/_app";

describe("MyApp", () => {
  test("renders snapshot of app component", () => {
    const wrapper = render(MyApp(<MyApp {...props} />));
    expect(wrapper).toMatchSnapshot();
  });
});
```

**Comment by thucne** [comment]
Hello, I was able to test the `_app.tsx` file as follows:

```tsx
import { render, screen, waitFor } from '@testing-library/react';
import App from '@/pages/_app';

describe('App', () => {
  it('_app.tsx test', () => {
    // Act
    render(<App Component={() => <div>Test</div>} pageProps={{}} router={{} as any} />);

    // Assert
    waitFor(() => {
      expect(screen.getByText('Test')).toBeInTheDocument();
    });
  });
});

```
